File: draft/notification_service.py
# api/services/notification_service.py
import logging
from ..models import NotificationConfig
from .nexah_service import NexahService
from .whatsapp_service import WhatsAppService

logger = logging.getLogger(__name__)

class NotificationService:
    """Service to handle sending notifications via configured channels"""
    
    @classmethod
    def send_otp(cls, recipient, otp_code, message=None):
        """
        Send OTP code using the configured default channel
        
        Args:
            recipient: Recipient phone number
            otp_code: The OTP code to send
            message: Optional custom message (used for SMS only)
        
        Returns:
            dict: Response with success status and details
        """
        # Get current configuration
        config = NotificationConfig.get_config()
        
        # Default SMS message if none provided
        if not message:
            message = f"Votre code de vérification Ngoma est : '{otp_code}'. Il expire dans 5 minutes."
        
        # Choose the appropriate service based on configuration
        if config.default_channel == 'whatsapp':
            logger.info(f"Sending OTP via WhatsApp to {recipient}")
            # Pour WhatsApp: ignore le message et utilise le template de la config
            return WhatsAppService.send_otp(
                recipient=recipient,
                otp_code=otp_code
            )
        else:  # Default to SMS
            logger.info(f"Sending OTP via SMS to {recipient}")
            # Pour SMS: utilise le message texte
            return NexahService.send_sms(
                recipient=recipient,
                message=message
            )

File: app/Services/Notifications/NotificationService.php
<?php

namespace App\Services\Notifications;

use App\Models\ServiceConfiguration;
use Illuminate\Support\Facades\Log;

class NotificationService
{
    protected WhatsAppService $whatsappService;
    protected NexahService $nexahService;

    public function __construct()
    {
        $this->whatsappService = new WhatsAppService();
        $this->nexahService = new NexahService();
    }

    /**
     * Send OTP code using the configured default channel
     *
     * @param string $recipient Recipient phone number
     * @param string $otpCode The OTP code to send
     * @param string|null $message Optional custom message (used for SMS only)
     * @return array Response with success status and details
     */
    public function sendOtp(
        string $recipient,
        string $otpCode,
        ?string $message = null
    ): array {
        // Get current configuration
        $defaultChannel = ServiceConfiguration::getDefaultNotificationChannel();

        // Default SMS message if none provided
        if (!$message) {
            $message = "Votre code de vérification Estuaire Emploie est : '{$otpCode}'. Il expire dans 5 minutes.";
        }

        // Choose the appropriate service based on configuration
        if ($defaultChannel === 'whatsapp') {
            Log::info("Sending OTP via WhatsApp to {$recipient}");

            $result = $this->whatsappService->sendOtp($recipient, $otpCode);

            // Fallback to SMS if WhatsApp fails
            if (!$result['success']) {
                Log::warning("WhatsApp failed, falling back to SMS");
                return $this->nexahService->sendSms($recipient, $message);
            }

            return $result;
        } else {
            // Default to SMS
            Log::info("Sending OTP via SMS to {$recipient}");

            $result = $this->nexahService->sendSms($recipient, $message);

            // Fallback to WhatsApp if SMS fails
            if (!$result['success']) {
                Log::warning("SMS failed, falling back to WhatsApp");
                return $this->whatsappService->sendOtp($recipient, $otpCode);
            }

            return $result;
        }
    }

    /**
     * Send notification via WhatsApp only
     *
     * @param string $recipient
     * @param string $otpCode
     * @return array
     */
    public function sendViaWhatsApp(string $recipient, string $otpCode): array
    {
        return $this->whatsappService->sendOtp($recipient, $otpCode);
    }

    /**
     * Send notification via SMS only
     *
     * @param string $recipient
     * @param string $message
     * @return array
     */
    public function sendViaSms(string $recipient, string $message): array
    {
        return $this->nexahService->sendSms($recipient, $message);
    }

    /**
     * Send a custom message (not OTP) via the default channel
     *
     * @param string $recipient
     * @param string $message
     * @return array
     */
    public function send(string $recipient, string $message): array
    {
        $defaultChannel = ServiceConfiguration::getDefaultNotificationChannel();

        if ($defaultChannel === 'whatsapp') {
            // For WhatsApp, we can only use templates, so we fallback to SMS for custom messages
            Log::info("Custom message requested, using SMS instead of WhatsApp (template limitation)");
            return $this->nexahService->sendSms($recipient, $message);
        } else {
            return $this->nexahService->sendSms($recipient, $message);
        }
    }
}
